pymc/tests/test_aesaraf.py

```diff
@@ -529,3 +529,23 @@ def test_rvs_to_value_vars():
     assert a_value_var in res_ancestors
     assert b_value_var in res_ancestors
     assert c_value_var in res_ancestors
+
+
+def test_rvs_to_value_vars_nested_transform_bug():
+    # See issue #5172
+    with pm.Model() as m:
+        one = pm.LogNormal("one", mu=0)
+        two = pm.LogNormal("two", mu=at.log(one))
+
+        # We add potentials or deterministics that are not in topological order
+        pm.Potential("two_pot", two)
+        pm.Potential("one_pot", one)
+
+        before = aesara.clone_replace(m.free_RVs)
+
+        # This call would change the model free_RVs in place
+        res, _ = rvs_to_value_vars(m.potentials, apply_transforms=True)
+
+        after = aesara.clone_replace(m.free_RVs)
+
+        assert equal_computations(before, after)
```

Comment:
I appreciate testing a reduced, more focused scenario, but it would be more reassuring to also test the full scenario that was initially observed.
That is doing `m.recompute_initial_point()` before and after a call to `pm.sample(chains=1, tune=1, draws=1)`.

It looks like you now understood much more of what's going on under the hood than I do, but when thinking and chatting about this issue last week we were also concerned about needing to maintain the `free_RVs` identity.
For example because they're keys in dicts like `m.initial_values`, or used as args to step methods. As far as I can tell the assert above checks only computation, not the identity?

Reply:
The identity was not changed. It was the inputs of (some of) free.RVs that were changed. This checks they are the same, which is a more strict check than what we were doing with the toposort (just checking if the number of nodes in the graph are the same). 

I can also check that the set of `graph_inputs` is exactly the same. That's what `recompute_initial_point` was complaining about.

Regarding the more complex test, I did check it locally and it's fixed as well. I prefer to have an example that is easier to understand. The case you found had a lot of moving parts that did not matter to raise the error. It was just the nested LogTransform out of topological order.

Reply:
Actually it seems `equal_computations` also makes sure the input variables are the same.